### ISSUE TITLE
one-var reports error for a variable declared in a for-loop after an uninitialized variable declaration

### ISSUE THREAD

**Issue by bajtos** [eslint/eslint#5744]
**What version of ESLint are you using?**
v2.5.3

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

```json
{
  "rules": {
    "one-var": ["error", {
      "initialized": "never",
      "uninitialized": "always"
    }]
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

Here is a small code snippet reproducing my issue. Just save it to a file and run `eslint` with the config shown above.

```js
function reproduce(candidates) {
  var err;
  for (var ix in candidates) {
    err = ix > 0;
  }
  console.log(err);
}
```

**What did you expect to happen?**

The code should pass linter checks.

**What actually happened? Please include the actual, raw output from ESLint.**

```
$ eslint bug.js

/private/tmp/bug/bug.js
  3:8  error  Combine this with the previous 'var' statement with uninitialized variables  one-var

✖ 1 problem (1 error, 0 warnings)
```

As far as my understanding goes, eslint considers the `ix` variable declared inside `for` loop as unitialized and therefore complains that it should have been defined on the same line as the `err` variable (before the `for` loop). Indeed, the following code passes the checks:

```js
function reproduce(candidates) {
  var err, ix;
  for (ix in candidates) {
    err = ix > 0;
  }
  console.log(err);
}
```

From my point of view, I am proposing one of the following two changes:
 - Fix the implementation of `one-var` to treat `for` loop variables differently.
 - Add an `one-var` option similar to `init-declaration'`s `ignoreForLoopInit` that will tell `one-var` to treat variables defined in for-loops differently.

Thoughts?

**Comment by bajtos** [comment]
A case that I find even more confusing:

```js
function printContextLogs(context) {
  for (var k in context.console._logs) {
    var items = context.console._logs[k];
    for (var ix in items) {
      console[k].apply(console, items[ix]);
    }
  }
}
```

The error message suggests merging the declaration of `ix` with `k`, which is rather nonsense to me.

**Comment by ilyavolodin** [comment]
@bajtos I assume the second example is done with the configuration as the first one? Because second example makes sense if configuration would've been:
```js
"one-var": ["error", {
  "initialized": "always",
  "uninitialized": "always"
}]
```
In that case, the example should be re-written as:
```js
function printContextLogs(context) {
  var k, ix;
  for (k in context.console._logs) {
    var items = context.console._logs[k];
    for (ix in items) {
      console[k].apply(console, items[ix]);
    }
  }
}
```

**Comment by bajtos** [comment]
@ilyavolodin thank you for coming back to me. Yes, the second example is using the same configuration as the first one.

Perhaps I should explain our coding style, here is an example:

```js
var a, b, c; // all uninitialized variables are in a single statement
var foo = 1;
var bar = 2; // initialized variables are one per statement.

// inside "for" loops, we consider variables as initialized
for (var i = 1; i < foo; i++) {
  for (var j = 1; j < bar; j++) {
    a = i*j;
  }
}
```

Thoughts?

**Comment by ilyavolodin** [comment]
Hmm... In that case, it sounds like we have a bug with uninitialized variables and loops. I was able to confirm this in the online demo. @bajtos do you want to submit a patch to fix this issue?

**Comment by bajtos** [comment]
> In that case, it sounds like we have a bug with uninitialized variables and loops. I was able to confirm this in the online demo. 

:+1: 

> do you want to submit a patch to fix this issue?

Uff, I am very tight on time, but I may give it a try, because I do want to get this fixed soon.

@ilyavolodin could you please point me in the direction where to start? Could it be this line, which treats differently only `ForStatement`, but not `ForInStatement` and `ForOfStatement`?

https://github.com/eslint/eslint/blob/7374d2b4b3838e6c14d2c45291687c25f5f7f191/lib/rules/one-var.js#L253

I assume the tests will be added to this file, is that correct?

https://github.com/eslint/eslint/blob/7374d2b4b3838e6c14d2c45291687c25f5f7f191/tests/lib/rules/one-var.js

**Comment by ilyavolodin** [comment]
@bajtos That looks like the right place. I verified that the rule doesn't warn for regular `for` loop statements, only for `for...in` and `for...of`, so it sounds like you just need to add those as special case along with `ForStatement`. And yes, tests would go into the file you linked.

**Comment by bajtos** [comment]
Ouch, I just realised I need to get an approval from my employer before I can contribute. It will take at least several days to get it processed. If there are any other volunteers to contribute the fix, then please don't hesitate to take this over.

**Comment by diki** [comment]
I would like to work on this, i think can submit the fix at weekend.

**Comment by ilyavolodin** [comment]
Sounds good @diki I think @bajtos outlined what needs to be done quite well in his previous post.

**Comment by diki** [comment]
Hi,

I want to clarify some points before going further.This code:

```js
var a, b, c; // all uninitialized variables are in a single statement
var foo = 1, bar = 2; // initialized variables are one per statement.

// inside "for" loops, we consider variables as initialized
for (var i = 1; i < foo; i++) {
  for (var j = 1; j < bar; j++) {
    a = i*j;
  }
}
``` 

fails with this configuration:

```js
module.exports = {
  rules: {
    'one-var': [2, {
      'initialized': 'never',
      'uninitialized': 'never'
    }],
  }
};
```
I think it is expectable, if so don't you think the same rule applies for uninitialized variables in ```ForInStatements```? Adding conditions to [one-var.js](https://github.com/eslint/eslint/blob/7374d2b4b3838e6c14d2c45291687c25f5f7f191/lib/rules/one-var.js) for uninitialized variables in ```ForInStatements``` would fix this, in the same manner i guess my example should pass too. Do you think ```ix``` in ```for(var ix in candidates)``` should be behaved as initialized?

**Comment by mysticatea** [comment]
I don't understand this thread completely, but I think variable declarations at `ForInStatement.left` and `ForOfStatement.left` should be treated as initialized. Those variables will be initialized with iteration items.
